compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmBuiltinCallLowering.kt

```diff
@@ -11,36 +11,64 @@ import org.jetbrains.kotlin.codegen.intrinsics.Not
 import org.jetbrains.kotlin.ir.declarations.IrFile
 import org.jetbrains.kotlin.ir.expressions.IrCall
 import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.impl.IrConstImpl
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.dump
+import org.jetbrains.kotlin.ir.util.isNullConst
+import org.jetbrains.kotlin.ir.util.isNullable
 import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
 import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
 
-// TODO: This lowering is currently JvmBackend specific because there are multiple
-// definitions of the boolean 'not' operator. Once there is only the irBuiltins
-// definition this lowering could be shared with other backends.
-class NegatedExpressionLowering(val context: JvmBackendContext) : FileLoweringPass {
+class JvmBuiltinCallLowering(val context: JvmBackendContext) : FileLoweringPass {
 
     companion object {
-        fun isNegation(expression: IrExpression, context: JvmBackendContext) : Boolean {
+        fun isNegation(expression: IrExpression, context: JvmBackendContext): Boolean {
             // TODO: there should be only one representation of the 'not' operator.
             return expression is IrCall &&
                     (expression.symbol == context.irBuiltIns.booleanNotSymbol ||
                             context.state.intrinsics.getIntrinsic(expression.symbol.descriptor) is Not)
         }
 
-        fun negationArgument(call: IrCall) : IrExpression {
+        fun negationArgument(call: IrCall): IrExpression {
             // TODO: there should be only one representation of the 'not' operator.
             // Once there is only the IR definition the negation argument will
             // always be a value argument and this method can be removed.
             return call.dispatchReceiver ?: call.getValueArgument(0)!!
         }
     }
 
+    private fun isNullCheckOfPrimitiveTypeValue(call: IrCall, context: JvmBackendContext): Boolean {
```

Comment:
Checked expression could have side effects:
calcPrimitiveWithSideEffects() != null

Reply:
Thanks for catching that Mike! I have updated the code to only allow the elimination of null checks for primitive type constants and for primitive type values loaded from a local variable. It is OK to eliminate the local variable load of a simple type value - even if it is changed during debugging it still cannot be updated to a null value. What do you think?

Reply:
As we discussed offline, in order to get rid of all null comparisons with primitive type arguments, we now transform the potentially side-effecting cases:

if (side-effecting-expr-with-primitive-type == null) ...

--->

side-effecting-expr-with-primitive-type
pop
if (false) ...